edx_rest_framework_extensions/tests/test_permissions.py

```diff
@@ -184,3 +185,114 @@ def test_has_permission(self, authentication_class, jwt_filters, view_kwargs, ex
+    )
+    @ddt.unpack
+    def test_has_permission(self, user_filters, requested_username, expected_result):
+        request = self._create_request(user_filters, requested_username)
+        self.assertEqual(
+            permissions.JwtHasUserFilterForRequestedUser().has_permission(request, None),
+            expected_result,
+        )
+
+
+class JwtRestrictionApplicationOrUserAccessTests(TestCase):
+    class SomeClassView(APIView):
+        authentication_classes = (SessionAuthentication, JwtAuthentication)
+        permission_classes = (permissions.JWT_RESTRICTED_APPLICATION_OR_USER_ACCESS,)
+
+        def get(self, request, course_id=None):
+            return Response(data="Success")
+
+    def _create_request(self, is_staff=False, username_in_url=False):
+        user = UserFactory(username='this_user', is_staff=is_staff)
+        url = '/'
+        if username_in_url:
+            url += '?username={}'.format(user.username)
+        request = RequestFactory().get(url)
+        request.user = user
+        request.auth = generate_jwt(request.user)
+        return request
+
+    def _create_session(self, request):
+        self.client.login(username=request.user.username)
+
+    def _create_jwt(self, request, is_restricted=False):
+        request.successful_authenticator = JwtAuthentication
+        request.auth = generate_jwt(request.user, is_restricted=is_restricted)
+
+    def _assert_in_log(self, text, mock_log_method):
+        self.assertTrue(mock_log_method.called)
+        self.assertIn(text, mock_log_method.call_args_list[0][0][0])
+
+    def test_anonymous_fails(self):
+        request = self._create_request()
+        response = self.SomeClassView().dispatch(request)
+        self.assertEqual(response.status_code, 403)
+
+    def test_session_staff_succeeds(self):
+        request = self._create_request(is_staff=True)
+        self._create_session(request)
+
+        response = self.SomeClassView().dispatch(request)
+        self.assertEqual(response.status_code, 200)
+
+    @patch('edx_rest_framework_extensions.permissions.log')
+    def test_session_user_not_in_url_fails(self, mock_log):
+        request = self._create_request()
+        self._create_session(request)
+        
+        response = self.SomeClassView().dispatch(request)
+        self.assertEqual(response.status_code, 403)
+        self._assert_in_log('IsUserInUrl', mock_log.info)
+
+    def test_session_user_in_url_succeeds(self):
+        request = self._create_request(username_in_url=True)
+        self._create_session(request)
+
+        response = self.SomeClassView().dispatch(request)
+        self.assertEqual(response.status_code, 200)
+
+    @patch('edx_rest_framework_extensions.permissions.waffle.switch_is_active')
+    def test_jwt_unrestricted_succeeds(self, waffle_mock):
+        waffle_mock.return_value = True
+        request = self._create_request(username_in_url=True)
+        self._create_jwt(request)
```

Comment:
If the jwt included is_restricted=True would this fail?  Should it?
Two separate questions related to my uncertainty around restricted. :)